### ISSUE TITLE
Thank you.

### ISSUE THREAD

**Issue by Granjow** [akhodakivskiy/VimFx#892]
@lydell and @akhodakivskiy – today, I had to switch to Vimium due to Firefox 57 (and other great extensions like Tab Groups and Beyond Australis got lost without replacement …).

Thanks for your work on this great extension. It really changed browsing from meh to fun. You are awesome.

**Comment by lydell** [comment]
Thank you! I really means a lot to hear things like this! :)

**Comment by mitermayer** [comment]
@lydell @akhodakivskiy interested if you guys have a rough idea on what would be required to allow support for firefox 57+. 

I may be willing to contribute to it

**Comment by lydell** [comment]
@mitermayer See #860. I think it is a better idea to contribute to [Saka Key](https://github.com/lusakasa/saka-key), though. Both projects are MIT licensed, so you can easily copy code from VimFx.

**Comment by sudoyang** [comment]
This is the best!. Thanks so much!

**Comment by ahstro** [comment]
Just wanted to add my thanks for VimFX. It was the first keyboard browsing plugin that worked exactly as I wanted it to. 🙏

**Comment by ainola** [comment]
After years of dealing with constantly broken pentadactyl/vimperator situations, I switched over to this fantastic extension. It was such a refreshing extension that balanced features and performance. I'm glad to have used it and am sad to see it go. Thanks so much, @lydell.

**Comment by levinit** [comment]
vimfx is practical.👍

**Comment by Nebucatnetzer** [comment]
I agree with all that was said before, VimFX was a really great addon and the first that really worked for me.

**Comment by jflorian** [comment]
Just want to chime in here as another very satisfied user that had started with pentadactyl then vimperator then a whole slew of others, including vimium before landing here -- it's always the last place you look!  This was the first one that I thought was done right all the way through.  Like a good friend passing, I wish I'd known it sooner.

I'd like to also mention that new Vim-like extensions are still popping up.  I just found [vim-vixen](https://github.com/ueokande/vim-vixen) but haven't had much time to evaluate it yet.  Seems very lean and clean.  Might be worth mentioning on the README here.

**Comment by Nebucatnetzer** [comment]
vim-vixen is quite nice but has the problem that you currently can't black list URLs. Beside that it's IMO the most promising one.

**Comment by jflorian** [comment]
@Nebucatnetzer , blacklist?  You mean like a way to turn off the extension for some sites?  Guess I've never used that feature on any of these, despite being a power user.  Guess it's just the sites I visit have never necessitated such a need.  If anything, I get upset when the key bindings aren't there as I'd expect, such as when looking at the Feedbro main page, which is more like a FF preferences page than a real web site.  Like what /proc is to a real filesystem.  Don't know what the FF parlance for that is.  Virtual page?

**Comment by bae22** [comment]
Just wanted to add my thanks to @lydell and @akhodakivskiy for all your work on VimFX. Can't imagine using Firefox without it, and sad to see it go.

**Comment by Nebucatnetzer** [comment]
@jflorian exactly yes. Sites like Google Docs for example have their own keybindings which then gets a bit hairy if they trigger before the vim plugins.

**Comment by girst** [comment]
From me too a big thank you to all the contributors and developers, especially @lydell and @akhodakivskiy of course! No other vim-type extension has put this much thought in the 'f'-link-labels -- the most important elements are actually the easiest combination of letters. 

Just a FYI: The addon still works in **Firefox Developer Edition 57** when you set `extensions.legacy.enabled` to true. (Same for Nightly 58, so this might be a viable path for the near future)

**Comment by Turysaz** [comment]
I also just wanted to say thank you.
My first vim-style-browsing experience was with Luakit, which has built-in vim-like keybindings. The UX of Luakit was good, but it often just did not feel polished.
When I then stumbled upon VimFx, it was like a revelation to me. VimFx has absolutely made my life a bit better. :-)
Thanks to all contributors for this great piece of software!

**Comment by john-soda** [comment]
Was the best add-on ever!
Tanks very much!
Can somebody could give me an advise, which one of these three add-ons is the closest to vim FX?
- Saka key
- Vimium-FF
- Vim Vixen
Thanks!

**Comment by ahstro** [comment]
@john-soda I don't know about "closest", but Vimium-FF has served my particular usage well. It does suffer from being a content script, so the page has to be loaded before it can be used, but I assume the others do as well (I don't remember that aspect of them). Saka and Vixen seemed to offer a few more "power user" features though.

**Comment by seltzy** [comment]
Thank you for this project. Sounds like it's dying now, but the past year of using it has been a blast. Much love to the maintainer and contributors. :)

**Comment by geniousli** [comment]
Thanks your great work, i have a pleasure time with your production

**Comment by anthony-geoghegan** [comment]
I've just upgraded to Firefox 57 and while I love the speed, my interaction with the browser UI feels crippled without VimFx (and ItsAllText). I was going to open a **Thank you** issue but it’s heartening to see that this one has been open since September.

Many thanks to @lydell @akhodakivskiy for providing such an elegant browser extension that has successfully avoided bloat and feature-creep to focus on providing a quality keyboard-based UI.

**Comment by spacepluk** [comment]
Another big **thank you** here, VimFX will be sorely missed!

**Comment by seltzy** [comment]
I've been evaluating alternatives and so far the biggest problem I have is the inability to select Firefox menu items and navigate the settings/addons pages. Granted VimFX was slightly buggy with this, but it did a good enough job for me. Does anyone know if this is impossible to implement using the new API, or is contributing the feature to other addons possible?

**Comment by lydell** [comment]
> select Firefox menu items

Not sure what you mean here?

>  navigate the settings/addons pages

WebExtensions are not allowed to do that.

> VimFX was slightly buggy with this

Buggy with what? Just curious 😉

**Comment by seltzy** [comment]
> Not sure what you mean here?

Whatever the `e b` command was called. It was nice because several extensions I used didn't have shortcut keys to open their menus, but they *did* have toolbar icons which could activate/focus their interface when clicked. Being able to "click" them via VimFX was *really* nice.

> Buggy with what? Just curious 😉

In that when I opened menu popups from "clicking" toolbar icons as described above, like the Firefox main menu, VimFX would not be available and the menu wouldn't focus. This meant I needed to use my mouse to click or focus on anything in it. This was an occasional pain that I learned to live with. I imagine this was a limitation of the extension API for security reasons, so I suppose "bug" was a misleading word.

A shame that I'll never be able to navigate options menus with keyboard alone. This is going to be a major annoyance for me going forward.

Still, this was a great addon. Hoping the new WebExtensions eventually get extended to support some of these features again.

**Comment by lydell** [comment]
I miss the `eb` command too (and, yes, the results did vary when it comes to what clickable things were found and how well they were activated). It is also one of the things that WebExtensions can't do.

**Comment by girst** [comment]
TIL about `eb`. I'll add my reminder that **VimFX still works on Firefox Developer Editon 58**.

**Comment by lydell** [comment]
@girst You mean the `extensions.legacy.enabled` pref, that only works in Developer Edition and Nightly? (For how long?) Not sure that's something we should promote. Dedicated users will find it anyway.

**Comment by girst** [comment]
yep, that's it. People are still looking for ways to work around the legacy addon deprecation. for how long? No guarantees, but I'll guess that most of the APIs VimFX uses will stay around for quite some time (and for me, every day I can use VimFX longer counts).

**Comment by pasau** [comment]
-rip- best addon in the history of addons

**Comment by seltzy** [comment]
Yeah, I'm seriously feeling the pain with it today. I've been putting together a [Saka Key](https://github.com/lusakasa/saka-key) configuration that mimics it since it seems to be the easiest alternative to retrofit, but it's just not the same... It has a few niceties that VimFX didn't, but it's missing a lot of features that I used on a daily basis. Like the next and previous when searching for text, or `a f` which was probably my favorite thing besides `e b`.

Guess some of us are just gonna have to pick an alternative to contribute to and hope things get better. I've never dabbled with extensions before, but it's probably not too hard. Like @lydell suggested, it may be best to extend [Saka Key](https://github.com/lusakasa/saka-key).

**Comment by jflorian** [comment]
@seltzy I missed it when I first tried Saka Key, maybe because it wasn't yet featured, but it now has a sense of "profiles" I believe is the term used where you can change all of the key bindings at once.  IIRC, one choice was to mimic VimFx.  I liked the look of the project but was totally turned off by the default bindings and diidn't want to mess around redoing them all.  It still won't be VimFx but if you didn't notice this setting it'll  be hard to adapt.

My biggest miss with Saka Key at this point is the ability to dictate which search engine I want by a keyword with `o` or `t`.  IIUC, that's what another extension, just Saka, is supposed to bring when it becomes available.

I tried giving up the whole VIM UI for Firefox and just adapt to the built-in key bindings since those always work regardless of it's a normal page or a built-in one, but the search engine/keyword thing is something I depend on so much.  If anyone knows who to do that with the regular address bar in plain old FF, please share.

**Comment by pasau** [comment]
for some reason vimfx was the only one that worked in about:newtab, about:addons, etc. it's annoying when you Shift+j a couple of tabs and get stuck in a tab where the shortcut won't work and you have to do Control+Tab on this one tab. Also the visual mode was perfect. i went back to vimium-ff, it's not as good.

**Comment by seltzy** [comment]
@jflorian I managed to get a basic profile working, but like you say, it needs some work to feel as smooth. As for `o`, you can use `ctrl+L`. Not as nice, but on my keyboard I'm usually using the edge of my palm to hit `ctrl` so it's only marginally more cumbersome. I'll bet there are other extensions out there you can use to re-map that.

@pasau I know, right? I wonder why they have problems with some websites. Seems like a good candidate for contribution. Probably just a matter of where/when to initialize the extension when the page loads.

**Comment by jflorian** [comment]
@seltzy I am aware of ctrl-l but not of a way to search using one of my many configured engines, only the default one. With the o command I can enter the search engine's keyword as a prefix. With engines as diverse as one for Python library documentation and another for Wikipedia, well you can imagine how that context helps.

**Comment by seltzy** [comment]
@jflorian Ah I see. I didn't know search engine configuration was a feature of VimFX. I use DuckDuckGo as my default and it has a [!bang](https://duckduckgo.com/bang) feature that can serve the same purpose. Not as direct, and it has the limitation of not being configurable on a per-user basis, but still it has served me well.

**Comment by lydell** [comment]
> Saka Key … one choice was to mimic VimFx.

The "VimFx profile" was very recently added to Saka Key, so I'm not sure if it has been released yet: https://github.com/lusakasa/saka-key/pull/98

> for some reason vimfx was the only one that worked in about:newtab, about:addons, etc.

There's a very good reason for this: WebExtensions are not allowed to access those tabs. So even if VimFx was updated, not even it could do it.

> they don't even work at all on addons.mozilla.org

That's some "security" thing Mozilla copied from Chrome extensions (which are not allowed to run on _their_ addons site).

> I didn't know search engine configuration was a feature of VimFX

Because there is none. I'm not sure what you're talking about.

**Comment by Mark-Eng** [comment]
Just want to add my thanks as well. I'm not a programmer and so had never heard of Vim or Vim-style keybindings before stumbling upon this add-on. Changed my life!

**Comment by spacepluk** [comment]
Another thing that annoys me about the alternatives is that none of them allow to map <C-n> to switch to the next tab :(

**Comment by lydell** [comment]
@spacepluk What do you mean? All of them have commands for switching tabs.

**Comment by spacepluk** [comment]
@lydell oops, bad markdown... I meant you can't map `<C-n>` *(edited above)*

**Comment by jflorian** [comment]
Responding here to my own mention of not knowing how to use FF's address bar to specify which search engine to use.  I just discovered, it's no different than Vimium and others.  You simply prefix the search with your search engine keyword.  So I have gg=Google, yt=YouTube, wp=Wikipedia, etc. which means I can do a Wikipedia search for, say vim with "wp vim".  As soon as I type "wp<space>" I see a helpful hint saying "Search with Wikipedia".

**Comment by seltzy** [comment]
@jflorian Yeah no idea what you're talking about. Maybe that's from another extension, or a feature of FF I've never stumbled across.

**Comment by jflorian** [comment]
@seltzy It's just a built-in feature.  Simply go to Preferences > Search and there you'll see One-Click Search Engines.  While they work I find adding them cumbersome.  It as is if the FF devs are trying to make them mysterious and magical.  Vimium has "custom search engines" which are independent but behave much the same way.  However, Vimium makes adding more trivial when an engine config is so simple as:

    wp: https://www.wikipedia.org/w/index.php?title=Special:Search&search=%s Wikipedia

The `%s` is the only magic bit; it merely gets replaced with your search terms.  So any site that has a search field that results in a URL just for that search is easily made into a "search engine" for FF.

VimFx got this right, like so many other aspects of Vim-ing FF -- let FF do things natively as much as possible.  I actually should have realized that VimFx wasn't doing anything special with these engines and then I wouldn't even had to raise my Q earlier regarding this.  My brain was tainted though by all the other Vim-ish extensions that did require me to set up search engines independent of FF.

So now I have my answer.  I just now need to decide which is better key bindings I don't like but which are consistent everywhere in FF or those that I much prefer but don't work on special pages meaning I kind of need to know two sets.  I suspect it'll be the latter because FF simply doesn't offer a way to follow links using the keyboard (that I know of).  One hopes that FF will eventually offer WebExtensions to work on any page.  IMHO this really should be up to the user.  Let me choose if I wish to trust an extension or not rather than blindly assuming it must be evil.  Why should I put more trust in FF than a particular extension?  (Don't get me wrong, I like being able to prevent some or even most from having such access.)

**Comment by seltzy** [comment]
I recommend lobbying for a more permissive WebExtension API so that a suitable VimFX replacement can be developed. It'd be nice if we could get Mozilla to adopt a smartphone-like permissions model so that you could grant certain extensions access to things like the `about:addons` tab or browser UI elements. I imagine things could move that way eventually.

I decided to punt and downgrade my home computer FF, accepting the risks. I'll keep Quantum on my work computer and explore Saka Key more. If I can make the time I'd like to contribute to it to get my most used features back. At least that way I can transition myself in a less disorienting fashion.

**Comment by lydell** [comment]
@seltzy The developer of Tridactyl is doing a great job working with Mozilla to come up with more keyboard-y APIs.

**Comment by seltzy** [comment]
@lydell Great to hear! Certainly will be keeping my eye open for that. Do you have a list of requests/reports that have been filed that people can comment on to draw more attention to them?

**Comment by lydell** [comment]
I think this is the main one: https://bugzilla.mozilla.org/show_bug.cgi?id=1215061

**Comment by emugel** [comment]
Both Saka and the new vimium are unusable on some pages, including most Mozilla Firefox pages. As soon as the address bar is selected I get stuck there. 

I will never ever use Google Chrome again, but until VimFx gets available, which I guess is never happening, I'll stick with palemoon.

**Comment by lydell** [comment]
@emugel WebExtensions are not allowed to run in `about:` pages as well as on `addons.mozilla.org`. There's currently no for WebExtensions way to focus the location bar, and as soon as you focus anything outside the current webpage WebExtensions are not able to run commands anymore. So even if VimFx was updated, it wouldn't help.

**Comment by sky-gh** [comment]
VimFx is the reason i purged firefox57+ and reinstalled and holded firefox56.
It works all the time, making it possible to work without a mouse.

**Comment by 8** [comment]
Thank you!
I had to switch to Waterfox to continue to use the plugin.
The other vim plugins are useless as they don't / can't implement even the most basic features.

Take care,
Martin

**Comment by Niekto350** [comment]
Agreed, the VimFx was the best Vim-like extension in the browser. I've tried many of the alternatives (if not all of them) and none came even close to VimFx. Shame that the mozilla architecture changes killed it just as I was starting to dig a bit deeper into it. It's the main reason why I've delayed switching to Firefox quantum. Big thank you to everyone who have been part of the project.

It's been a few years since then and I still come back here from time to time in secret hopes that it would get resurrected. :)

**Comment by girst** [comment]
On Sat, Feb 19, 2022 at 01:59:48PM -0800, Niekto350 wrote:
>It's been a few years since then and I still come back here from time 
>to time in secret hopes that it would get resurrected. :)

erm, it has! you'll need to install the legacyfox shim (see readme § 
installation), and it'll work with current firefox releases.

**Comment by Niekto350** [comment]
> On Sat, Feb 19, 2022 at 01:59:48PM -0800, Niekto350 wrote: It's been a few years since then and I still come back here from time to time in secret hopes that it would get resurrected. :)
> erm, it has! you'll need to install the legacyfox shim (see readme § installation), and it'll work with current firefox releases.


Thank you, I've seen that in the past but never got that working. I'll have to take a closer look at it again when it's confirmed to be working good.

Just one more question, does it work even on the "protected" pages like a new tab as it did before?

**Comment by girst** [comment]
On Sat, Mar 05, 2022 at 02:11:44AM -0800, Niekto350 wrote:
>Just one more question, does it work even on the "protected" pages like 
>a new tab as it did before?

yes.

**Comment by slalik** [comment]
Thank you! Thank you so much! VimFX is what makes FF useful!

I find the vimfx interface even more enjoyable than in qutebrowser due to the more balanced switching between normal and insert modes.

I'm so happy!
